Add RandomSampler

arcondello opened the pull request:
Add a performant random sampler. This is intended to replace `dimod.RandomSampler` as part of https://github.com/dwavesystems/dimod/issues/298.

This implements the same API, but also includes a `time_limit` parameter. This is a feature used by the benchmarking team - this sampler is meant to replace their random batch sampler with one maintained as part of Ocean.

I have also attempted to include timing information following the spec in https://github.com/dwavesystems/dwave-samplers/issues/22.

I kind of hate that, because we list the samplers alphabetically, this one gets put first :shrug:.

arcondello commented:
This relies on an unreleased dimod feature. Will open a PR shortly.

kevinchern commented on README.rst:
```diff
@@ -24,13 +24,32 @@ or locally on your CPU.
 *dwave-samplers* implements the following classical algorithms for solving
 :term:`binary quadratic model`\ s (BQM):
 
+* Random: a sampler that draws uniform random samples.
```
Missing backtick symbol i.e, 
```
* `Random` : a sampler that draws uniform random samples.
```

arcondello replied:
Those are for the link. In this case I am not linking to anything. Though open to a good place to link to, https://en.wikipedia.org/wiki/Randomization seemed a bit over general :smile:

JoelPasvolsky replied:
It's backtick plus underscore to link to the new Random section in line 35

arcondello replied:
Got it, I misunderstood. Will do.

kevinchern commented on README.rst:
```diff
@@ -24,13 +24,32 @@ or locally on your CPU.
+Random
+======
+
+Random samplers provide a useful baseline performance comparison. The variable
```
Would this be the place to mention which PRNG is used? e.g., Mersenne-Twister?

arcondello replied:
Could do, though the user can actually set that by passing in a [Generator](https://numpy.org/doc/stable/reference/random/generator.html#numpy.random.Generator). By default we use NumPy's default, which is currently [PCG64](https://numpy.org/doc/stable/reference/random/bit_generators/pcg64.html#numpy.random.PCG64). Though I would want to just say that we use the default from NumPy in case NumPy changes.

kevinchern commented on dwave/samplers/random/cyrandom.pyx:
```diff
@@ -0,0 +1,174 @@
+    cdef double sampling_start_time = realtime_clock()
+
+    cdef double sampling_stop_time
+    if time_limit < 0:
```
Should this be `<=`?

kevinchern commented on dwave/samplers/random/cyrandom.pyx:
```diff
@@ -0,0 +1,174 @@
+    for i in range(cybqm.num_variables()):
+        state.sample.push_back(bitgen.next_uint32(bitgen.state) % 2)
+
+    if is_spin:
+        # go back through and convert to spin
+        for i in range(state.sample.size()):
+            state.sample[i] = 2 * state.sample[i] - 1
+
+    state.energy = cybqm.data().energy(state.sample.begin())
```
Converting spins and computing energies seem more fitting as a postprocessing step.

arcondello replied:
Only in the case where you're keeping every sample. Otherwise you need the energies in order to keep the population to `num_reads`. Because for "real uses" of the solver you need to keep the population limited, IMO it's part of sampling.

arcondello replied:
Four different combos with their uses:
- both unspecified: I just want to use the defaults
- `time_limit` set, `num_reads` unset: I want the best sample I can find randomly in the given time
- `time_limit` unset, `num_reads` set: I want `num_reads` samples drawn uniformly. If I was to specify the time_limit, they would not be uniform.
- both set, I want the best sample(s) found after `time_limit` and I want to specify the number of returned samples and/or the internal memory used.

Obviously these are not inclusive of use cases but should give some idea.

randomir replied:
I would forbid user setting both `num_reads` and `time_limit`, since `time_limit` takes precedence and might return less than `num_reads`.

arcondello replied:
https://github.com/dwavesystems/dwave-samplers/pull/24/commits/947320b1a28522c278a4520ad4182c17ed7f05ab changed the behavior to always return `num_reads`. `time_limit` now just allows the user to keep sampling.

randomir replied:
I was looking at your comment above, and missed that update, thanks.

kevinchern commented on tests/test_random_sampler.py:
```diff
@@ -0,0 +1,64 @@
+        sampleset = RandomSampler().sample(bqm, num_reads=10, time_limit=.02)
+        runtime = time.perf_counter() - t
+
+        self.assertTrue(.01 < runtime < .04)
```
Should also test info['timing']['sampling_time'] is reasonably close to 0.02.

kevinchern commented on tests/test_random_sampler.py:
```diff
@@ -0,0 +1,64 @@
+        self.assertTrue(.01 < runtime < .04)
+
+        dimod.testing.assert_sampleset_energies(sampleset, bqm)
+        self.assertEqual(len(sampleset), 10)
```
also `info['num_drawn']`

arcondello commented on dwave/samplers/random/cyrandom.pyx:
```diff
@@ -0,0 +1,174 @@
+
+    sampleset.info.update(
+        num_drawn=num_drawn,
+        prepreocessing_time=sampling_start_time-preprocessing_start_time,
+        sampling_time=postprocessing_start_time-sampling_start_time,
+        postprocessing_time=realtime_clock()-preprocessing_start_time,
```
I am open to having it be nested. Though since we're already changing the names, IMO consistency is not the main reason to do it.

arcondello replied:
If we do nest it, and if we're already changing the names, IMO it would look nicer to have
```python
info = dict(timing=dict(
    preprocessing=...,
    sampling=...,
    postprocessing=...,
))
```

CatherineCMcGeoch replied:
IMHO consistency of timing terminology with QPU and other solvers is a priority.  Otherwise the code to plot data from different solvers on the same graph becomes very unwieldy.   This may require some exceptions and compromises because different solvers structures have different structures, but aiming for maximum comparability would be good thing.   This would also be an argument for generic timing categories based on ``once per input'' and ``once per output'' ,  which are natural comparators, rather than solver-specific functions.

kevinchern replied:
+1 for nested timings!

randomir replied:
I would also consider typing the duration variables. Perhaps best to use `datetime.timedelta`. I always need to check if some time variable in Ocean is in seconds, ms or us. Hate that.

randomir replied:
QPU timing info could be considered encoded for transport -- and we can decode it as `timedelta` client-side.

arcondello commented on dwave/samplers/random/sampler.py:
```diff
@@ -0,0 +1,122 @@
+            time_limit:
+                The maximum sampling time in seconds.
+                If given and non-negative, samples are drawn until ``time_limit``.
+                Only the best ``num_reads`` (or fewer) samples are kept.
+
+        Returns:
+            A sample set.
+            Some additional information is provided in the
+            :attr:`~dimod.SampleSet.info` dictionary:
+
+                * **num_drawn**: The total number of samples generated.
```
Specifically when `time_limit` is provided and when it allows for more reads

kevinchern commented on tests/test_random_sampler.py:
```diff
@@ -0,0 +1,64 @@
+        with self.assertWarns(dimod.exceptions.SamplerUnknownArgWarning):
+            RandomSampler().sample(bqm, a=5, b=2)
+
+    def test_time_limit(self):
```
Maybe add test cases for different settings:
1. default (no `num_reads`, no `time_limit`)
2. only `num_reads`
3. only `time_limit`

arcondello replied:
The default case is covered rather extensively by `@dimod.testing.load_sampler_bqm_tests(RandomSampler)`

kevinchern reviewed (commented):
Maybe add a few more test cases, otherwise it looks good.

randomir commented on dwave/samplers/random/cyrandom.pyx:
```diff
@@ -0,0 +1,154 @@
+        while realtime_clock() < sampling_stop_time:
+
+            samples.push_back(get_sample(cybqm, bitgen, is_spin))
+            sort(samples.begin(), samples.end(), comp_state)
+            samples.pop_back()
```
Perhaps use a heap instead? (for total `O(N logN)` complexity)

Alternatively, you can use insertion sort (for `O(N)` per step here, and total `~O(N^2)`) instead of "IntroSort" (for `O(N logN)` per step here, and total of `~O(N^2 logN)`).

randomir replied:
Or constant factor too big to notice benefits on smaller scale.

arcondello replied:
Ok, after the change introduced in https://github.com/dwavesystems/dwave-samplers/pull/24/commits/176c25c2e8d86ecf3b51161482248c35644b570f, I was able to get a performance boost at very large `num_reads`. So I will indeed make the change. Thanks, and good suggestion!

arcondello replied:
For posterity I used
```python
bqm = dimod.generators.gnp_random_bqm(1000, .5, 'BINARY')
num_reads=10000
```

randomir commented on dwave/samplers/random/sampler.py:
```diff
@@ -0,0 +1,122 @@
+                Seed for the random number generator.
+                Passed to :func:`numpy.random.default_rng()`.
+
+            time_limit:
+                The maximum sampling time in seconds.
+                If given and non-negative, samples are drawn until ``time_limit``.
+                Only the best ``num_reads`` (or fewer) samples are kept.
```
Until we figure out how best to extend/generalize `Initialized` sampler interface to handle other termination criteria, I would forbid setting both `num_reads` and `time_limit` here. It's confusing because `time_limit` takes precedence and the sampler might return less then `num_reads`.

A case of setting `time_limit`, but not `num_reads` is also confusing -- we return only one sample after `num_drawn` tries.

Perhaps we should introduce `max_answers` (or `max_samples`) that can be used in combination with `time_limit`. See https://github.com/dwavesystems/dwave-greedy/issues/24 for discussion.

So:
- `num_reads` set: exactly `num_reads` returned, `time_limit` not allowed. `max_answers` ignored.
- `time_limit` set: `max_answers` (which can default to 1 or `inf`/None) is used

With this scheme, I would set `num_reads` by default to `None`, with interpretation on how much that actually is depending on `time_limit`/`max_answers`. Similarly to how we resolve it in the `Initialized` arg parser.

arcondello replied:
I had already changed it so `num_reads` has precedence. And I disagree that they should be mutually exclusive. The solver will now always return `num_reads`. `time_limit` just determines whether it keeps going.

randomir replied:
Ok, so it looks an implied AND between `==num_reads` and `>=time_limit`:
- sampler will always return `num_reads`,
- sampler will run for at least `time_limit`?

I find this awkward and inconsistent with `Initialized` (`num_reads` adapts to `initial_states` if states are user-specified and reads are not) and termination criteria in dwave-hybrid/`Kerberos` (time/convergence/energy/iteration count are OR-ed - we stop as soon as one condition is met).

arcondello replied:
I don't think it needs to be consistent with `Initialized` because it doesn't have any concept of an initial state.

My understanding from benchmarking is that they prefer to give the sampler a fixed runtime, so we need some sort of time-based exit criteria.

So, we can either do treat `time_limit` as the most important, and then have a `max_answers`/`max_samples` as the secondary criteria. Or we can treat `num_reads` and the most important and treat `time_limit` as the secondary criteria when provided.

I much prefer the latter for a few reasons:
1) I don't know how to choose a good default `time_limit`, since the amount of solving will vary a lot for different BQMs.
2) for even moderately long time limits, the `num_reads` will be a tiny fraction. So while it is possible for `time_limit` to be exceeded for large `num_reads`, most typical uses will not run into that. It's also addressed by making `num_reads` default 1. 
3) `num_reads` is consistent with most of our other samples.

randomir replied:
I find disjunction (OR) between termination criteria much simpler and cleaner.

- If user sets `num_reads` (and not `time_limit`), they obviously don't care about run time, and just want to get specific number of samples back.

- If user sets `time_limit` (and not `num_reads`), they care about run time, although it's not clear what's the number of samples they want. But if they haven't specified the number, it makes sense to return _all_ samples generated during runtime. Here's where `max_samples`/`max_answers` comes into play.

- In this framework, setting both `num_reads` and `time_limit` would run until one of them is satisfied.

arcondello replied:
Ok, how about I just make two samplers. One that take a `time_limit` and `max_answers` and one that takes `num_reads`. If we're going to make the arguments disjoint, we might as well make that clear at the sampler level.

Personally I think that's way overkill, but it seems like no one else agrees with me :smile:

arcondello replied:
> If user sets time_limit (and not num_reads), they care about run time, although it's not clear what's the number of samples they want. But if they haven't specified the number, it makes sense to return all samples generated during runtime. Here's where max_samples/max_answers comes into play.

The problem is that we generate a lot of samples. Way more than can easily fit in memory. So returning _all_ is not practical for many problems.

> In this framework, setting both num_reads and time_limit would run until one of them is satisfied.

This is what I had before.

kevinchern replied:
> So, we can either do treat time_limit as the most important, and then have a max_answers/max_samples as the secondary criteria. Or we can treat num_reads and the most important and treat time_limit as the secondary criteria when provided.

If we're deciding between the two, I also favour the latter for similar reasons. I do think Radomir's proposal is more intuitive and gives more control or the behaviour of the sampler.

Another point to consider is if we add a similar feature for Neal, a consistent behaviour across classes of samplers would be desirable. And in the case of Neal, I think it's reasonable to set a time limit retrieve all answers it came across.

randomir replied:
Extending sampler interface with `time_limit` (and `max_answers` or whatever we call it) would be a good candidate for a follow-up PR, just to keep concerns separated. But since we had all this discussion here, we might as well do it here and then generalize in a follow-up.

randomir commented on releasenotes/notes/random-sampler-a5bacc2b7761222f.yaml:
```diff
@@ -0,0 +1,3 @@
+---
+features:
+  - Add ``RandomSampler`` which generates samples with variable assignments chosen by coin flip.
```
Unless we have a coin flipping device attached to user's machine, I would rephrase this :smile:

randomir reviewed (commented):
Looks good overall, my only concern is about sampling termination criteria interface.

arcondello commented:
Ok, refactored. Now both `num_reads` and `time_limit` are termination criteria. Added `max_num_samples` to limit the population of samples kept internally.

I also have dropped the timing info, pending a decision on #22. We can restore it in a followup PR.

randomir commented on README.rst:
```diff
@@ -24,13 +24,32 @@ or locally on your CPU.
+
+Get the best 5 sample found in .1 seconds
+
+>>> sampleset = sampler.sample(bqm, time_limit=.1, max_num_samples=5)
```
Perhaps include ~`.info['num_drawn']`~ `.info['num_reads']` here to clarify many more were actually generated.

randomir commented on dwave/samplers/random/sampler.py:
```diff
@@ -0,0 +1,160 @@
+                The maximum number of random samples to be drawn.
+                If neither ``num_reads`` nor ``time_limit`` are provided,
+                ``num_reads`` is set to 1.
+                If ``time_limit`` is provided, there is no limit imposed on
+                the number of reads.
```
Isn't the limit set by `max_num_samples` then?

arcondello replied:
That's the limit on the population, it will take unlimited reads up to the time limit.
